cotix/_collision_resolution.py

```diff
@@ -13,68 +13,253 @@
-    #  probably based on the body properties
-    # todo: angular momentum
+    contact_point_speed1 = (
+        jnp.dot(body1.velocity, change_of_basis[1])
+        + relative_contact_point1[0] * body1.angular_velocity
+    )
+    contact_point_speed2 = (
+        jnp.dot(body2.velocity, change_of_basis[1])
+        + relative_contact_point2[0] * body2.angular_velocity
+    )
+    tangential_relative_velocity = contact_point_speed1 - contact_point_speed2
+    return lever_arm1, lever_arm2, tangential_relative_velocity
 
-    elasticity = 1
+
+def _resolve_collision(
+    body1: AbstractBody, body2: AbstractBody, contact_info: ContactInfo
+) -> Tuple[AbstractBody, AbstractBody, CollisionResolutionExtraInfo]:
+    """
+    Resolves a collision between two bodies.
+    Potentially changes velocities, positions and angular velocities of the bodies.
+    Contains a check that closest points of the bodies are moving apart.
+      Does nothing if they are.
+    """
+    penetration_vector = contact_info.penetration_vector
+    contact_point = contact_info.contact_point
+    elasticity = body1.elasticity * body2.elasticity
+    friction_coefficient = body1.friction_coefficient * body2.friction_coefficient
 
     # change coordinate system from (x, y) to (q, r)
-    # where q is the line along epa_vector and r is perpendicular to it
-    unit_collision_vector = epa_vector / jnp.linalg.norm(epa_vector)
+    # where [0] is the line along epa_vector and [1] is perpendicular to it
+    unit_collision_vector = penetration_vector / jnp.linalg.norm(penetration_vector)
     perpendicular = perpendicular_vector(unit_collision_vector)
     change_of_basis = jnp.array([unit_collision_vector, perpendicular])
     change_of_basis_inv = jnp.linalg.inv(change_of_basis)
 
-    v1q = jnp.dot(body1.velocity, unit_collision_vector)
-    v1r = jnp.dot(body1.velocity, perpendicular)  # stays constant
-    v2q = jnp.dot(body2.velocity, unit_collision_vector)
-    v2r = jnp.dot(body2.velocity, perpendicular)  # stays constant
+    # everything below should be in the new coordinate system
+    v1_col_basis = change_of_basis @ body1.velocity
+    v2_col_basis = change_of_basis @ body2.velocity
 
-    v1q_new, v2q_new = _1d_elastic_collision_velocities(
-        body1.mass, body2.mass, v1q, v2q
+    v_rel = v1_col_basis - v2_col_basis
+
+    (
+        lever_arm1,
+        lever_arm2,
+        tangential_relative_velocity,
+    ) = _compute_lever_arms_and_tangential_velocity(
+        change_of_basis, contact_point, body1, body2
     )
 
-    v1qr_new = jnp.array([v1q_new * elasticity, v1r])
-    v2qr_new = jnp.array([v2q_new * elasticity, v2r])
+    # check that the collision has to be resolved
+    point1_v = body1.velocity + body1.angular_velocity * lever_arm1
+    point2_v = body2.velocity + body2.angular_velocity * lever_arm2
+    # todo: this might not work with super concave bodies,
+    #  when the center of mass is on the wrong side of the collision point
```

Comment:
i think  this is not important for now

Reply:
Yeah whatever, agreed

Reply:
Though.... We probably should think at some point how to do it correctly. Like an example of fully concave body would be a body with a big hole inside, where some objects are moving. While this is not the case for most simulation, this is a plausible scenario. Anyway, that's the thing for the future us to care

Reply:
yep, thats why i left it as a todo for now